torch/distributed/distributed_c10d.py

```diff
@@ -761,12 +761,26 @@ def _new_process_group_helper(
             assert backend.upper() in Backend._plugins, (
                 f"unknown c10d backend type {backend.upper()}"
             )
-            pg = Backend._plugins[backend.upper()](
-                prefix_store, rank, world_size, timeout
+
+            backend_creator_fn = Backend._plugins[backend.upper()]
+
+            if pg_options is None:
+                import inspect
+                backend_module = inspect.getmodule(backend_creator_fn)
+                assert 'Options' in dict(inspect.getmembers(backend_module, inspect.isclass)), (
+                    f"Process group Options not defined in {backend.upper()} implementation"
+                )
+
+                pg_options = backend_module.Options.create()
```

Comment:
Why do we need all of this? Can't we just pass in `None` to `backend_creator_fn` instead?

Reply:
The group options have to be a custom class extending from the `ProcessGroup::Options`.  In the case of 3rd party extensions, this requires adding a pybind definition like the following. 

```
py::class_<c10d::ProcessGroupTest::Options, c10::intrusive_ptr<c10d::ProcessGroupTest::Options>>(m, "Options")
            .def(py::init<const bool &>())
            .def("create", &c10d::ProcessGroupTest::Options::create)
            .def_readwrite("some_test_property", &c10d::ProcessGroupTest::Options::some_test_property);
```

Passing None in this cause pybind to fail not being able to create an instance of the custom options type

Reply:
We should leave for the backend creation function the to fill in with a default options object.

This eliminates this BC breaking requirement that there must be a class named `Option` in the same module as the creation function.

Reply:
This has been fixed and will be pushed in the next iteration